tracer/src/Datadog.Trace/TraceContext.cs

```diff
@@ -63,17 +73,18 @@ public void AddSpan(Span span)
 
                     if (_samplingPriority == null)
                     {
-                        if (span.Context.Parent is SpanContext context && context.SamplingPriority != null)
+                        if (span.Context.Parent is SpanContext { SamplingPriority: { } samplingPriority })
                         {
                             // this is a root span created from a propagated context that contains a sampling priority.
-                            // lock sampling priority when a span is started from a propagated trace.
-                            _samplingPriority = context.SamplingPriority;
+                            // no need to track sampling mechanism since we won't override the propagated tag/header.
+                            _samplingPriority = samplingPriority;
+                            _samplingMechanism = null;
                         }
-                        else
+                        else if (Tracer.Sampler is not null)
                         {
                             // this is a local root span (i.e. not propagated).
-                            // determine an initial sampling priority for this trace, but don't lock it yet
-                            _samplingPriority = Tracer.Sampler?.GetSamplingPriority(RootSpan);
+                            // determine an initial sampling priority for this trace and track the sampling mechanism.
+                            (_samplingPriority, _samplingMechanism) = Tracer.Sampler.MakeSamplingDecision(span);
```

Comment:
Should we call `AddSamplingMechanismTag` here? Should the root span always be sending this value even if it was an auto decision by the agent?

Reply:
Uhm, yes. We need the mechanism tag to be present in case we make downstream calls, just like sampling priority. Waiting until the end of the trace will not work. Thanks!

Reply:
Thanks. I moved things around. We now set both sampling priority and mechanism when adding the first span to the trace context.